### ISSUE TITLE
[Bug]: OffloadingConnector can silently return wrong output at exact chunk boundaries with mamba_cache_mode=all

### ISSUE THREAD

**Issue by jairitAge** [vllm-project/vllm#51094]
### Your current environment

Reproduced with:

- vLLM: `0.25.1` nightly, commit `752a3a504`
- Model: `nvidia/NVIDIA-Nemotron-Nano-9B-v2` (hybrid full-attention + Mamba)
- `enable_prefix_caching=True`
- `mamba_cache_mode="all"` (the model's default when prefix caching is enabled)
- `OffloadingConnector`, CPU KV offload enabled
- eager execution, greedy decoding

The relevant condition is still present in v0.26.0 and current `main` (`43c4bdcae9e589fde99984d1d5d03cbdeecf716b` as checked on 2026-08-05), although I have not rerun the GPU reproduction on those revisions yet.

### Describe the bug

`OffloadingConnector` can silently produce different output after a CPU-cache refetch when the prompt length is exactly a multiple of the offloaded Mamba chunk size and `mamba_cache_mode="all"`.

`resolve_mamba_align_size()` currently enables hit-window alignment only for `mamba_cache_mode == "align"`:

```python
if isinstance(kv_spec, MambaSpec) and kv_spec.mamba_cache_mode == "align":
    ...
```

However, `"all"` mode also stores recurrent Mamba states at token/block positions. A prefix lookup is capped at `num_tokens - 1` because the final token must be recomputed for logits. At an exact chunk boundary, without alignment the offload lookup can therefore return `N - 1` tokens while restoring the state stored at boundary `N`. That state already contains token `N`, so recomputing token `N` from it applies the token twice and changes the output without raising an exception.

### Reproduction

1. Create an `LLM` with the configuration above.
2. Use a deterministic token prompt whose length is exactly two offload chunks. In this model/configuration the chunk size was 1408 tokens, so the prompt length was 2816.
3. Generate once with greedy decoding to populate the CPU offload cache.
4. Clear the GPU prefix cache while retaining the connector cache (`reset_connector=False`; on the tested revision the default reset retained it).
5. Generate the identical prompt again.

Observed:

```text
cold:    prompt=2816 cached=0    first tokens=[18631, 43085, 17091, 47695, 1556]
refetch: prompt=2816 cached=2815 first tokens=[18631, 43085, 17091, 47695, 1270]
```

The engine completes normally, but the fifth generated token differs.

With the one-line change below, the refetch is rounded back to a valid recurrent-state boundary (`cached=1408` in this setup) and the generated output is bitwise identical to the cold run:

```diff
- if isinstance(kv_spec, MambaSpec) and kv_spec.mamba_cache_mode == "align":
+ if isinstance(kv_spec, MambaSpec) and kv_spec.mamba_cache_mode in ("align", "all"):
```

### Expected behavior

A CPU-offloaded prefix hit must never resume from a recurrent state that already includes the token being recomputed. Cold execution and CPU-refetch execution should produce identical greedy output.

### Suggested regression test

The existing `test_mamba_align_cpu_offload` already tests exact-boundary and mid-block prompts, but explicitly sets `mamba_cache_mode="align"`. Parameterizing the boundary case with `"all"` should cover this failure.

### Related but separate issue

#46453 tracks per-group prefix-hit divergence for hybrid Mamba models. This report is specifically about offload hit-window alignment in `mamba_cache_mode="all"` at an exact chunk boundary.

**Comment by maithilijoshi20** [comment]
Hi, I’d like to work on resolving this issue and will raise a PR with a fix and regression coverage.
